Login page

vbaltrusaitis-reef opened the pull request:
This PR implements a login page. The flow is this:

1. Login page shows a simple email and password form.
2. When a user submits the login form, a fetch call is fired to the authentication endpoint.
3. If the authentication call succeeds, the user is redirected either to the URL that was passed on via `?next=` parameter or (if there was no such parameter) to `settings.LOGIN_REDIRECT_URL`.

By default, this in no way interferes with other login mechanisms. You have to change `settings.LOGIN_URL` in order to use this as a default login page.

Since this PR adds a view that is not an JSON API view, I've also changed the URL schema to separate pages from API endpoints. If this doesn't seem right, just let me know.

A note about `DJANGO_FIREBASE_AUTH_FALLBACK_LOGIN_URL` settings parameter. This parameter sets a fallback login URL in case Firebase isn't working. I was thinking of using `settings.LOGIN_URL` for this by default at first, but the problem is that the user of this library may want to use Firebase login page as the `settings.LOGIN_URL` page. This would result in a circular fallback link that just links to itself. It would be possible to check whether Firebase login view is set as `LOGIN_URL` and use `LOGIN_URL` as a fallback only if it's not (using [django.urls.reverse()](https://docs.djangoproject.com/en/4.0/ref/urlresolvers/#reverse)), but I'm not quite sure this is a good idea.

vbaltrusaitis-reef committed:
Add login page

vbaltrusaitis-reef committed:
Use submit form instead of a button

vbaltrusaitis-reef commented on django_firebase_auth/views.py:
```diff
@@ -70,7 +82,7 @@ def _get_or_create_user(email: str) -> Optional[AbstractBaseUser]:
             return None
 
-    user = UserModel.objects.create_user(email=email, is_active=True)
+    user = UserModel.objects.create_user(username=email, email=email, is_active=True)
```
`create_user` requires a username at least with the default `django.contrib.auth`.
@mpnowacki-reef did you have some custom auth module in mind?
Using `email` as a username may be fine for some use-cases, but maybe not all. Should I make this behavior configurable? I.e. use either email or firebase ID, or something else based on settings?

mpnowacki-reef replied:
yeah, the one project we use this app in currently uses a custom UserModel with no `username` field.

mpnowacki-reef replied:
although I think this additional property will just be ignored

mpnowacki-reef replied:
now that I think of it, as a next extension of this app, a function for creating users should be configurable

mpnowacki-reef replied:
no, it breaks our current app. I think for now let's leave it without `username` and later on let's add that configurable hook for creating users.

vbaltrusaitis-reef replied:
Ok, removed the `username` parameter.

vbaltrusaitis-reef committed:
Remove unnecessary print()

mpnowacki-reef commented on django_firebase_auth/views.py:
```diff
@@ -92,3 +104,19 @@ def _verify_firebase_account(headers: HttpHeaders) -> str:
+
+
+def login_page(request: HttpRequest) -> HttpResponse:
+    if request.user.is_authenticated:
+        return redirect(request.GET.get('next', settings.LOGIN_REDIRECT_URL))
+    template = loader.get_template('firebase_authentication/login.html')
+    return HttpResponse(
+        template.render({
+            'firebase_web_api_key': WEB_API_KEY,
+            'jwt_header_name': JWT_HEADER_NAME,
+            'firebase_auth_endpoint': reverse(authenticate),
+            'login_redirect_url': settings.LOGIN_REDIRECT_URL,
+            'django_login_url': settings.LOGIN_URL,
+            'fallback_login_url': FALLBACK_LOGIN_URL,
+        })
+    )
```
```suggestion
next = request.GET.get('next', reverse('admin:index'))
    if request.user.is_authenticated:
        return redirect(next)
    template = loader.get_template('firebase_authentication/login.html')
    return HttpResponse(
        template.render({
            'firebase_web_api_key': WEB_API_KEY,
            'jwt_header_name': JWT_HEADER_NAME,
            'firebase_auth_endpoint': reverse(authenticate),
            'login_redirect_url': next,
            'django_login_url': settings.LOGIN_URL,
            'fallback_login_url': FALLBACK_LOGIN_URL,
        })
    )
```

vbaltrusaitis-reef replied:
Using `admin:index` now. Also passing proper `next` parameter.

vbaltrusaitis-reef committed:
Fix: allow parameters to be named URL patterns

Before the change, redirects and URLs would only work if parameters
LOGIN_REDIRECT_URL, LOGIN_URL and
DJANG_FIREBASE_AUTH_FALLBACK_LOGIN_URL
were actual URL as opposed to Django's names URL patterns.
Using resolve_url fixes this.

vbaltrusaitis-reef committed:
Create users without username

This app will -- for now -- asume that we're using a custom UserModel
which does not require a username parameter.

vbaltrusaitis-reef committed:
Add a button to login with Django user

Also removes the fallback login URL.

vbaltrusaitis-reef committed:
Allow redirection to 404 pages

Before this commit, the redirection would result in server error,
even though just redirecting to non-existing page is appropriate.

vbaltrusaitis-reef commented:
The last few commits implement both Firebase and Django login in one page. Same fields for email and password are used. There are two buttons - one for login with Firebase and one for login with Django. By default (i.e. if you press enter) Firebase login is used.

mpnowacki-reef commented on django_firebase_auth/views.py:
```diff
@@ -92,3 +108,63 @@ def _verify_firebase_account(headers: HttpHeaders) -> str:
+                return self._render(request, next, "Wrong password")
+
+        except UserModel.DoesNotExist:
+            return self._render(request, next, "Wrong email")
```
```suggestion
            user: AbstractBaseUser = UserModel.objects.get(email=email)
        except UserModel.DoesNotExist:
            return self._render(request, next, "Wrong email")
        if user.check_password(password):
            login(request, user=user, backend=AUTH_BACKEND)
            return redirect(next)
        else:
            return self._render(request, next, "Wrong password")

```

vbaltrusaitis-reef replied:
Resolved by [ff3adc5](https://github.com/reef-technologies/django-firebase-auth/pull/2/commits/ff3adc583c96f75615f97ef6cbf5910e558f64d2).

vbaltrusaitis-reef committed:
Show error when user does not belong to staff

mpnowacki-reef commented on django_firebase_auth/views.py:
```diff
@@ -162,7 +169,10 @@ def post(self, request: HttpRequest) -> HttpResponse:
                 login(request, user=user, backend=AUTH_BACKEND)
-                return redirect(next)
+                if request.user.is_staff:
+                    redirect(next)
```
```suggestion
                    return redirect(next)
```

vbaltrusaitis-reef committed:
Fix a bug: missing return statement

Co-authored-by: Michał Nowacki <77497496+mpnowacki-reef@users.noreply.github.com>

vbaltrusaitis-reef committed:
Fix: catch exception closer to the source

vbaltrusaitis-reef committed:
Use the same message for bad email or password
